File: DS in C or C++/Stack Implementations and Application/stack.cpp
#include<iostream>
using namespace std;
class Stack {
private:
	int *stack;
	int top;
	int list_capacity;
	/*Method for increasing the list capacity*/
	void increase_capacity(int amount) {
		if (amount < 0) {
			amount = 5;
		}
		int *temp = new int[list_capacity = list_capacity + amount];
		for (int i = 0;i < (top + 1) ; i++) {
			*(temp + i) = *(stack + i);
		}
		delete[] stack;
		stack = temp;
	}
public:

	/* Constructors for Stack */
	Stack() {
		stack = new int[list_capacity = 10];
		top = -1;
	}

	Stack(int init_capacity) {
		if (init_capacity < 2) {
			stack = new int[list_capacity = 10];
			top = -1;
		}
		else {
			stack = new int[list_capacity = init_capacity];
			top = -1;
		}
	}

	/* Utility Methods for Stack*/
	void push(int item) {
		if (top != list_capacity -1) {
			*(stack + ++top) = item;
		}
		else {
			increase_capacity(10);
			*(stack + ++top) = item;
		}
	}

	int pop() {
		if (top != -1) {
			return *(stack + top--);
		}
		else {
			cout<<"Stack is Empty"<<endl;
			return -1;
		}
	}

	int size() {
		return top + 1;
	}

	int peek() {
		if (top!=-1) {
			return *(stack + top);
		}
		else {
			cout<<"Stack is empty"<<endl;
			return -1;
		}
	}

	void empty() {
		top = -1;
	}
};
int main() {
	Stack obj(45);
	obj.push(44);
	obj.push(3);
	cout<<"size of stack is : "<<obj.size();
	cout<<"Pop : "<<obj.pop()<<endl;
	cout<<"size of stack is : "<<obj.size();
	obj.empty();
	cout<<"size of stack is : "<<obj.size();
	return 0;
}

File: DS in Java/Stack/Stack.java
/**
 * Data Structures : Stack
 * Filename : stack.java
 * @author Sachin Singh
 * Email : sachinsingh.sk13@gmail.com
 * Description : basic stack class with push() , pop() , peek() and size() methods.
 * */

public class Stack<T> {
    Object[] stack;
    int tos;
    int capacity;

    /*increase stack capacity by 50% */
    private void increase_capacity() {
        Object[] temp = new Object[capacity = capacity + capacity/2];
        for (int i = 0;i < stack.length ;i++) {
            temp[i] = stack[i];
        }
        stack = temp;
    }

    public Stack() {
        stack = new Object[10];
        tos = -1;
        capacity = 10;
    }

    public Stack(int init_capacity) {
        tos = -1;
        if (init_capacity>2) {
            stack = new Object[init_capacity];
            capacity = init_capacity;
        }
        else {
            stack = new Object[10];
            capacity = 10;
        }
    }

    public void push(T item) {
        if (tos != capacity - 1) {
            stack[++tos] = item;
        }
        else {
            increase_capacity();
            stack[++tos] = item;
        }
    }

    public T pop() {
        if (tos != -1) {
            return (T) stack[tos--];
        }
        else {
            return null;
        }
    }

    public T peek() {
        if (tos != -1) {
            return (T) stack[tos];
        }
        else {
            return null;
        }
    }

    public int size() {
        return tos+1;
    }

    public void empty() {
        stack = null;
        // reinitialize with capacity 10
        stack = new Object[10];
        capacity = 10;
        tos = -1;
    }
}
